### DISCUSSION TITLE
RX 580 and Linux: SteamVR Home and Beat Saber are laggy, display driver crashes, jitter?

### DISCUSSION THREAD

**Discussion by Espionage724** [alvr-org/ALVR#860] [Q&A]
I'm of the understanding that ALVR on Linux works great with AMD GPUs, but I can't get it working ideally at all with my RX 580. When starting SteamVR, ALVR either crashes my display driver, or in the rare occasions it does work, FPS is low and everything (including SteamVR Home and Beat Saber) is jittery and laggy.

Other people in Discord make ALVR out to be usable with other AMD GPUs, but I'm unsure why I can't get it figured out with a RX 580.

#### Environment:

- CPU: Ryzen 7 2700X
- GPU: RX 580 (8GB, POLARIS10, DRM 3.42.0, 5.14.15-300.fc35.x86_64, LLVM 13.0.0) (0x67df))
- Audio: Realtek ALC 1120 (ASUS PRIME-X470 PRO built-in, 4.0 speaker channels)
- Headset: Quest 2 (V35)
- OS: Fedora 35 (GNOME 41)
- Kernel: 5.14.15-300.fc35.x86_64
- glxinfo: [Pastebin](https://pastebin.com/xZT25XhH)
- ALVR: alvr-16.0.0-1.0.0+1635915618+8dc973b6.x86_64 (trae32566)
- SteamVR: 1.20 and linux_v1.14.16

#### Description

- I start Steam (desktop launcher entry), and then launch ALVR (desktop launcher entry), which then automatically starts up SteamVR
- Dashboard is accessible from http://localhost:8082/ with Firefox (Chrome isn't installed)
- Usually when launching ALVR on my Quest 2, it will say it's restarting the server, freeze the screen on the headset, and then my computer's display will go black, show some purple corruption all over everything for maybe 5-10 seconds, and then blank my display entirely.
- In the rare-event ALVR launches properly on both the headset and computer, SteamVR Home starts, and is jittery and low FPS. I launch Beat Saber on the PC and it appears in the HMD, but it too is a jittery low FPS mess. [Here is a short video showing the low FPS and slow drawing on the sides over USB](https://www.youtube.com/watch?v=Bg2I2xmG55A).
- I've had this happen on Fedora 34, openSUSE Tumbleweed, with the pre-compiled Ubuntu binaries on the Nightly repo, self-compiled, and trae32566's repo

#### Troubleshooting

- Tried different wireless routers (known-good with Air Link on Windows)
- Did USB forwarding ([Option 2 - SideQuest](https://github.com/alvr-org/ALVR/wiki/Use-ALVR-through-a-USB-connection#option-2---sidequest)
- Tried Xorg and Wayland on GNOME 40 and 41
- Tried stable, beta, and linux_1.14 SteamVR builds
- Did `setenforce 0` on Fedora 35 before starting Steam and ALVR
- Verified that all known PCVR streaming apps (Virtual Desktop, ALVR, Radeon ReLive VR, wired and Air Link) are usable from Windows on this computer with expected performance in both H.264 and H.265

**Comment by Espionage724** [comment]
This seems to happen to enough people to make this an unlikely reason, but I wonder if the drive being used for ALVR and/or SteamVR and its sector size would affect anything with this?

**TLDR**: If your HDD or SSD either came with a native 4K sector mode enabled, or you've manually enabled it on your own, try changing it to 512e mode instead. **It's worth verifying this through multiple tools and product spec pages if you're at all unsure about this.** Windows 10 has a command, and Linux can display it clearly with `hdparm` and `nvme-cli`.

___

I notice a few minor issues for the past few months after switching my NVMe SSD to native 4K sector size, but never associated those issues to the sector size. Everything installed and ran fine as far as I knew, but possibly the most interesting non-issue was how my racing wheel pedals bound with odd axis names (like instead of Gas/Brake/Clutch, it'd be axiz-x, axis-y, and the Clutch binding being named "Gas"); this happened both with G HUB from Windows and new-lg4ff on Linux. My BIOS would fail all tests on the NVMe controller and namespace, but I had just casually chalked that up to the BIOS being odd or not directly supporting my SSD fully and assumed nothing was wrong. I sometimes noticed Steam games I restored too

The peak of this came when I recently changed my GPU from a RX 580 to a RTX 3060 ([thread on Guru3D](https://forums.guru3d.com/threads/just-got-a-3060.441092/)); had major installation issues with Steam and Oculus, and was completely clueless as to why. [Someone on Steam forums](https://steamcommunity.com/discussions/forum/1/4739473745776804012/?tscn=1641033020#c3203747342915419588) reported they solved their issue by going from 4K to 512e for the sector size, and knowing that I went out of my way to switch to 4K sectors, I entertained that idea and sure enough everything that I've tried after that point worked flawlessly (Steam installed, Oculus updated, pedal names bound with their actual Gas/Brake/Clutch names, NVMe test passed in BIOS).

As to the actual issue, I have no idea. I would assume my SSD doesn't properly handle 4K sectors (Sk hynix P31 1TB), but this doesn't quite explain why the major issues only happened when going from AMD to an NVIDIA GPU. It's possible this could be related to some kind of update starting around mid-December 2021 (Windows 10 and 11 both got new CUs around the 12th). Outside of those few things I tested that were obvious, there's no telling what other issues may have been caused by that, and since the wheel axis thing happened in Linux too, I imagine this isn't limited to Windows (although interesting LTSC 2019 was fine at quick-glance, presumably because it isn't as up-to-date).

I changed the lbaf format from `0` (512e) to `1` (4K) with `nvme-cli format` from Linux, and then simply put it back to `0` to go back to 512e ([some rough notes here](https://wiki.realmofespionage.xyz/notes:drive_wipe#sector_size)). This will require a full drive erase and OS reinstall. Presumably this process varies between drive models. For the P31, I had to suspend the system and wake it to unlock it. To be somewhat fair, I don't notice anything that advertises the sector size for a P31 and it's possibly there's known-issues for Hynix to not be setting it by-default, but interestingly 4K sectors is officially advertised in the firmware as a lbaf format.
___

I don't know when or if I'll be able to test ALVR on Linux with my RX 580 again any time soon, but this may be something interesting for anyone else to look at for any troubleshooting. I have not tested personally if ALVR was affected in any manner from this.

I would have likely caught this as the direct cause earlier if major problems came up shortly after doing the change, but I was fine for about 4 months, and the major problems only came after switching a GPU. All drive and file integrity tests from Windows and Linux passed no problem, and the only slight hint I got that there was a drive-related issue was that Steam and Oculus both reported a NTFS file transaction error in deep logs (which I had chalked-up to just those programs being broken in some manner).

The point I'm trying to really get across here is that I changed a drive's sector size to native 4K thinking it was better, ran fine with it for months, ran into unexpected major issues only after changing a GPU, didn't think to circle-back to the drive sector size as a reason (because it worked fine prior to that), and only got lucky someone else mentioned this solving their issue. Maybe this will hopefully help another person in this situation!
